cmd/tsidp/tsidp.go

```diff
@@ -1243,3 +1276,80 @@ func isFunnelRequest(r *http.Request) bool {
+// It needs to have permissions to get, update and patch the Secret.
+// If the Secret does not already exist, it also needs to have permissions to create it.
+func validateKubePermissions(state string) error {
+	secretName, ok := strings.CutPrefix(state, "kube:")
+	if !ok || secretName == "" {
+		return fmt.Errorf("unable to retrieve valid Kubernetes Secret name from %q", state)
+	}
+
+	kc, err := kubeclient.New("tailscale-tsidp")
+	if err != nil {
+		return fmt.Errorf("error initializing kube client: %w", err)
+	}
+
+	// Our kube client connects to kube API server via the kubernetes
+	// Service in the default namespace, which is not the default client-go
+	// etc behaviour and causes issues to some users. The client defaults
+	// probably cannot be changed for backwards compatibility reasons, but
+	// we can do the right thing here at the same time as adding support for
+	// tsidp to be deployed to kube.
+	url, err := kubeAPIServerAddress()
+	if err != nil {
+		return fmt.Errorf("error initiating kube client: %w", err)
+	}
+	kc.SetURL(url)
+
+	// CheckSecretPermissions returns an error if the permissions to get or
+	// update the Secret are missing. We also require that patch permission
+	// is present and, if the Secret does not exist, create permission.
+	canPatch, canCreate, err := kc.CheckSecretPermissions(context.Background(), secretName)
+	if err != nil {
+		return fmt.Errorf("error checking permissions for Secret %q: %w", secretName, err)
+	}
+	_, err = kc.GetSecret(context.Background(), secretName)
+	if err != nil && kubeclient.IsNotFoundErr(err) && !canCreate {
+		return fmt.Errorf("kube state Secret %s does not exist and we do not have permissions to create it, ensure that the right RBAC has been applied", secretName)
+	} else if err != nil && !kubeclient.IsNotFoundErr(err) {
+		return fmt.Errorf("error attempting to get kube state Secret %q: %w", secretName, err)
+	}
+	// tailscaled state updates are written to the state Secret by patch
+	// operations, so we must have permissions to patch it.
+	if !canPatch {
+		return fmt.Errorf("we do not have the required permissions to patch the state Secret %q, ensure that the right RBAC has been applied", secretName)
+	}
+	return nil
+}
+
+// kubeAPIServerAddress determines the address of the kube API server. It uses
+// the standard environment variables set by kube that are expected to be found
+// on any Pod- this is the same logic as used by client-go.
+// https://github.com/kubernetes/client-go/blob/v0.29.5/rest/config.go#L516-L536
+func kubeAPIServerAddress() (_ string, err error) {
+	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+	if host == "" {
+		err = errors.New("[unexpected] tsidp seems to be running in a Kubernetes environment with KUBERNETES_SERVICE_HOST unset")
+	}
+	if port == "" {
+		newErr := errors.New("[unexpected] tsidp appears to be running in a Kubernetes environment with KUBERNETES_SERVICE_PORT unset")
+		if err != nil {
```

Comment:
this is more complicated than needed. You can just do:

    err = multierr.New(err, errors.New("...."))

And the multierr.New constuctor will ignore the first err if it's nil and return the second one unwrapped.